### PR TITLE
Add a method to retrieve a list of available resources for IResourceStores

### PR DESCRIPTION
**Pull request by nyquillerium**
* Closes https://github.com/ppy/osu-framework/issues/2436
* [x] Depends on https://github.com/ppy/osu-framework/pull/2462

### CODE REVIEW

File: osu.Framework/IO/Stores/StorageBackedResourceStore.cs [hunk 1]
```diff
@@ -46,6 +47,19 @@ public virtual async Task<byte[]> GetAsync(string name)
 
         public Stream GetStream(string name) => storage.GetStream(name);
 
+        public IEnumerable<string> GetAvailableResources()
```
Comment (peppy) [inline osu.Framework/IO/Stores/StorageBackedResourceStore.cs]: no test to test this

File: osu.Framework/IO/Stores/ResourceStore.cs [hunk 2]
```diff
@@ -185,6 +185,20 @@ public void AddExtension(string extension)
                 searchExtensions.Add(extension);
         }
 
+        public virtual IEnumerable<string> GetAvailableResources()
+        {
+            lock (stores)
+            {
+                foreach (var store in stores)
+                {
+                    var contents = store.GetAvailableResources()?.ToList() ?? new List<string>();
```
Comment (peppy) [inline osu.Framework/IO/Stores/ResourceStore.cs]: why is the `new List` required

File: osu.Framework/IO/Stores/NamespacedResourceStore.cs [hunk 3]
```diff
@@ -20,6 +21,8 @@ public NamespacedResourceStore(IResourceStore<T> store, string ns)
             Namespace = ns;
         }
 
+        public override IEnumerable<string> GetAvailableResources() => base.GetAvailableResources().Select(name => $"{Namespace}/{name}");
```
Comment (peppy) [inline osu.Framework/IO/Stores/NamespacedResourceStore.cs]: Why is the namespace included here? isn't this store abstracting it away?

File: osu.Framework/IO/Stores/CachedResourceStore.cs [hunk 4]
```diff
@@ -104,6 +105,8 @@ public void ResetCache()
             cache.Clear();
         }
 
+        public override IEnumerable<string> GetAvailableResources() => base.GetAvailableResources().Concat(cache.Keys);
```
Comment (peppy) [inline osu.Framework/IO/Stores/CachedResourceStore.cs]: Why add cache keys? Isn't this going to cause duplicate entries?

File: osu.Framework/Graphics/Textures/TextureStore.cs [hunk 5]
```diff
@@ -85,5 +86,7 @@ public new virtual Texture Get(string name)
                 return tex;
             }
         }
+
+        public override IEnumerable<string> GetAvailableResources() => base.GetAvailableResources().Concat(textureCache.Keys);
```
Comment (peppy) [inline osu.Framework/Graphics/Textures/TextureStore.cs]: Why concat cache here?

File: osu.Framework/Graphics/Textures/PrefixTextureStore.cs [hunk 6]
```diff
@@ -16,5 +18,7 @@ public PrefixTextureStore(string prefix, IResourceStore<TextureUpload> stores)
         }
 
         public override Texture Get(string name) => base.Get($@"{prefix}-{name}");
+
+        public override IEnumerable<string> GetAvailableResources() => base.GetAvailableResources().Select(name => $"{prefix}-{name}");
```
Comment (peppy) [inline osu.Framework/Graphics/Textures/PrefixTextureStore.cs]: why add the prefix? this store is abstracting away the prefix.

File: osu.Framework/Audio/Sample/SampleManager.cs [hunk 7]
```diff
@@ -66,5 +67,7 @@ protected override void UpdateState()
         }
 
         public Stream GetStream(string name) => store.GetStream(name);
+
+        public IEnumerable<string> GetAvailableResources() => sampleCache.Keys;
```
Comment (peppy) [inline osu.Framework/Audio/Sample/SampleManager.cs]: shouldn't this implementation be delegated to `store`? else you are only returning previously accessed items.

**Comment by peppy** [comment]
The return values of `GetAvailableResources()` should be immediately usable in respective store's `Get` method. The way you've implemented this for prefix/namespace stores is opposite to that.

File: osu.Framework/IO/Stores/ResourceStore.cs [hunk 8]
```diff
@@ -185,6 +185,20 @@ public void AddExtension(string extension)
...
 
+        public virtual IEnumerable<string> GetAvailableResources()
+        {
+            lock (stores)
+            {
+                foreach (var store in stores)
+                {
+                    var contents = store.GetAvailableResources();
+
+                    foreach (string str in contents.Select(x => $"{store}/{x}"))
```
Comment (smoogipoo) [inline osu.Framework/IO/Stores/ResourceStore.cs]: Why is `store` being prefixed here?

**Comment by smoogipoo** [comment]
Please check your testcases. They make no sense.

File: osu.Framework/IO/Stores/ResourceStore.cs (hunk 8 above)
Comment (nyquillerium) [reply osu.Framework/IO/Stores/ResourceStore.cs]: Was originally trying to prefix fontname on glyphstores (since thats how you would perform the `Get()`), but that was incorrect here anyway. 

I added the fontname as a prefix in GlyphStore. If there's a problem with that let me know.

Comment (smoogipoo) [reply osu.Framework/IO/Stores/ResourceStore.cs]: That was what I was trying to point out, that it's incorrect here. It's fine at `GlyphStore`.

File: osu.Framework/IO/Stores/StorageBackedResourceStore.cs (hunk 1 above)
Comment (peppy) [inline osu.Framework/IO/Stores/StorageBackedResourceStore.cs]: pointless literal string `$"{file}"` (this likely wouldn't get optimised out at compile-time). also could rewrite in a more concise way:

```chsarp
public IEnumerable<string> GetAvailableResources() =>
            storage.GetDirectories(string.Empty).SelectMany(storage.GetFiles);
```

File: osu.Framework/IO/Stores/ResourceStore.cs [hunk 9]
```diff
@@ -185,6 +185,20 @@ public void AddExtension(string extension)
                 searchExtensions.Add(extension);
         }
 
+        public virtual IEnumerable<string> GetAvailableResources()
```
Comment (peppy) [inline osu.Framework/IO/Stores/ResourceStore.cs]:
```csharp
public virtual IEnumerable<string> GetAvailableResources()
        {
            lock (stores) return stores.SelectMany(s => s.GetAvailableResources());
        }
```

File: osu.Framework.Tests/Visual/Platform/TestSceneResourceStores.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            };
+
+            this.fontStore = fontStore;
+            this.storage = storage;
+        }
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddStep("Clear text container", () => textContainer.Clear());
```
Comment (peppy) [inline osu.Framework.Tests/Visual/Platform/TestSceneResourceStores.cs]: i'd probably just do this in `populateText`

File: osu.Framework.Tests/Visual/Platform/TestSceneResourceStores.cs [hunk 11]
```diff
@@ -18,9 +19,10 @@ public class TestSceneResourceStores : FrameworkTestScene
         private FontStore fontStore;
         private FillFlowContainer textContainer;
         private Storage storage;
+        private TextureStore textureStore;
 
         [BackgroundDependencyLoader]
-        private void load(FontStore fontStore, Storage storage)
+        private void load(FontStore fontStore, Storage storage, AudioManager audioManager, TextureStore textureStore)
```
Comment (peppy) [inline osu.Framework.Tests/Visual/Platform/TestSceneResourceStores.cs]: audiomanager????

File: osu.Framework/Graphics/Textures/PrefixTextureStore.cs [hunk 12]
```diff
@@ -16,5 +19,7 @@ public PrefixTextureStore(string prefix, IResourceStore<TextureUpload> stores)
         }
 
         public override Texture Get(string name) => base.Get($@"{prefix}-{name}");
+
+        public override IEnumerable<string> GetAvailableResources() => base.GetAvailableResources().Where(x => Regex.IsMatch(x, $@".*{prefix}-.*"));
```
Comment (peppy) [inline osu.Framework/Graphics/Textures/PrefixTextureStore.cs]: this will match more than you expect. i can easily pull a failcase from my ass.

**Comment by peppy** [comment]
I strongly suggest you add tests for the remaining stores, and have the tests actually perform a `Get` on the store so you can be sure they are working as expected. You could even add actual asserts and make them real tests.

It would also make sense to test each store which has different logic implemented for `GetAvailableResaources` to attain full test coverage.

File: osu.Framework/Graphics/Textures/PrefixTextureStore.cs (hunk 12 above)
Comment (nyquillerium) [reply osu.Framework/Graphics/Textures/PrefixTextureStore.cs]: This was actually already changed locally, pending testing. Blocking for now while thats happening.

File: osu.Framework/Game.cs [hunk 13]
```diff
@@ -124,11 +124,11 @@ private void load(FrameworkConfigManager config)
             Textures.AddStore(Host.CreateTextureLoaderStore(new OnlineStore()));
             dependencies.Cache(Textures);
 
-            var tracks = new ResourceStore<byte[]>(Resources);
+            var tracks = new ResourceStore<byte[]>();
```
Comment (peppy) [inline osu.Framework/Game.cs]: this needs to be in own pr

**Comment by peppy** [comment]
rewrote tests
